core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -1878,6 +1878,98 @@ class SocketServerTest {
+      processRequest(testableServer.dataPlaneRequestChannel, request1)
+      receiveResponse(sslSocket)
+
+      // then put 2 requests in SslTransportLayer.netReadBuffer via the ProxyServer
+      val connectionId = request1.context.connectionId
+      val listener = testableServer.config.dataPlaneListeners.head.listenerName.value
+      val channel = testableServer.dataPlaneAcceptor(listener).get.processors(0).channel(connectionId).getOrElse(throw new IllegalStateException("Channel not found"))
+      val transportLayer: SslTransportLayer = JTestUtils.fieldValue(channel, classOf[KafkaChannel], "transportLayer")
+      val netReadBuffer: ByteBuffer = JTestUtils.fieldValue(transportLayer, classOf[SslTransportLayer], "netReadBuffer")
+
+      proxyServer.enableBuffering(netReadBuffer)
+      sendRequest(sslSocket, requestBytes)
+      sendRequest(sslSocket, requestBytes)
+
+      val keysWithBufferedRead: util.Set[SelectionKey] = JTestUtils.fieldValue(testableSelector, classOf[Selector], "keysWithBufferedRead")
+      keysWithBufferedRead.add(channel.selectionKey)
+      JTestUtils.setFieldValue(transportLayer, "hasBytesBuffered", true)
+
+      // process the first request in the server side
+      // this would move bytes from netReadBuffer to appReadBuffer, then process only the first request
+      // we call wakeup() so Selector.poll() does not block in this step (because we artificially add data into netReadBuffer)
+      testableSelector.wakeup()
+      val req1 = receiveRequest(testableServer.dataPlaneRequestChannel, selectTimeout + 1000)
+      processRequest(testableServer.dataPlaneRequestChannel, req1)
+
+      // receive response in the client side
+      receiveResponse(sslSocket)
+
+      // process the second request in the server side
+      // this would process the second request in the appReadBuffer
+      // NOTE 1: this should not block because the data is already in the buffer, but without the fix for KAFKA-13559,
+      // this step will block for 300 ms (in this test, we override poll() timeout to 5000 ms to make it distinct)
+      // NOTE 2: we do not call wakeup() here so Selector.poll() would block if the fix is not in place
+      val processTimeStart = System.nanoTime()  // using nanoTime() because it is meant to calculate elapsed time
+      val req2 = receiveRequest(testableServer.dataPlaneRequestChannel, selectTimeout + 1000)
+      processRequest(testableServer.dataPlaneRequestChannel, req2)
+      val processTimeEnd = System.nanoTime()
+
+      // receive response in the client side
+      receiveResponse(sslSocket)
+
+      // check the duration of processing the second request
+      val processTimeDuration = (processTimeEnd - processTimeStart) / 1000000.0  // in ms
+      assertTrue(processTimeDuration < selectTimeout,
+        "Time to process the second request (" + processTimeDuration + " ms) should be under " + selectTimeout + " ms")
+
```

Comment:
likewise, we can probably remove this code.

Reply:
I assume you are referring to `receiveResponse()` call. I have removed it as well.